sdk/iot/pnp/src/az_iot_pnp_client.c

```diff
@@ -53,32 +54,42 @@ AZ_NODISCARD az_result az_iot_pnp_client_init(
 
 AZ_NODISCARD az_result az_iot_pnp_client_get_user_name(
     az_iot_pnp_client const* client,
-    az_span mqtt_user_name,
-    az_span* out_mqtt_user_name)
+    char* mqtt_user_name,
+    size_t mqtt_user_name_size,
+    size_t* out_mqtt_user_name_length)
 {
   AZ_PRECONDITION_NOT_NULL(client);
-  AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
+  AZ_PRECONDITION_NOT_NULL(mqtt_user_name);
+  AZ_PRECONDITION(mqtt_user_name_size > 0);
 
-  az_span remainder = mqtt_user_name;
-  az_span user_name;
-  AZ_RETURN_IF_FAILED(
-      az_iot_hub_client_user_name_get(&client->_internal.iot_hub_client, remainder, &user_name));
+  size_t written;
 
-  int32_t written = az_span_size(user_name);
-  remainder = az_span_slice_to_end(remainder, written);
+  // First get hub user name since it is unknown how long it will be
+  AZ_RETURN_IF_FAILED(az_iot_hub_client_user_name_get(
+      &client->_internal.iot_hub_client, mqtt_user_name, mqtt_user_name_size, &written));
 
-  int32_t required_length
-      = az_span_size(pnp_model_id) + az_span_size(client->_internal.root_interface_name) + 2;
+  az_span mqtt_user_name_span
```

Comment:
Nit: I think the code is easier to read if you do the length checks before creating the span, which may get not used if the `AZ_RETURN_IF_NOT_ENOUGH_SIZE` isn't satisfied.

Something like
```c
int required_length = ...;
AZ_WHATEVER_ASSERT_SHOULD_BE(required_length > mqtt_user_name_size);
az_span mqtt_user_name_span = ... // we'll know it'll be OK
```
```

Reply:
I would agree but currently `AZ_RETURN_IF_NOT_ENOUGH_SIZE` requires the first param to be a span. I can alternatively just have something like:

```c
if (required_size > mqtt_user_name_size)
{
    return AZ_ERROR_INSUFFICIENT_SIZE;
}
```
and do away with the macro.

Reply:
I think I will resolve this one and say that if we decide we want to do this (and I can ask the team), then we can do it across all the API's instead of just these.

Reply:
I do agree we should be consistent and I wouldn't block checkin trying to resolve this style issue.  

But I do think your `if` block up higher in comments is way easier to read even if more verbose, because we're checking before we're assigning.  Maybe (eventually) a macro that just deals with lengths and bringing in consistency?

Reply:
A length based macro seems like a reasonable addition. Same with just doing the explicit length check and not having a macro. The macro was more useful when span had 3 fields, because the required size check was a bit more complicated. Although, it is useful to have a centralized place that returns the `AZ_ERROR_INSUFFICIENT_SIZE` so if we were to change that, it only needs to be updated in one place.

Reply:
It would need to be changed at the macro but the arguments to it would need to be update too so slightly more intensive. But yes I agree. We can think about changing the macro since the check based on lengths/size numbers seems like it would be more appropriate